test/common/grpc/google_async_client_cache_test.cc

```diff
@@ -0,0 +1,91 @@
+#include "common/grpc/google_async_client_cache.h"
+
+#include "test/mocks/server/factory_context.h"
+
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+
+namespace Envoy {
+namespace Grpc {
+namespace {
+
+class AsyncClientCacheTest : public testing::Test {
+public:
+  AsyncClientCacheTest() {
+    client_cache_singleton_ = std::make_unique<AsyncClientCacheSingleton>();
+  }
+
+  void expectClientCreation() {
+    factory_ = new Grpc::MockAsyncClientFactory;
+    async_client_ = new Grpc::MockAsyncClient;
+    EXPECT_CALL(async_client_manager_, factoryForGrpcService(_, _, true))
+        .WillOnce(Invoke([this](const envoy::config::core::v3::GrpcService&, Stats::Scope&, bool) {
+          EXPECT_CALL(*factory_, create()).WillOnce(Invoke([this] {
+            return Grpc::RawAsyncClientPtr{async_client_};
+          }));
+          return Grpc::AsyncClientFactoryPtr{factory_};
+        }));
+  }
+
+  NiceMock<ThreadLocal::MockInstance> tls_;
+  Grpc::MockAsyncClientManager async_client_manager_;
+  Grpc::MockAsyncClient* async_client_ = nullptr;
+  Grpc::MockAsyncClientFactory* factory_ = nullptr;
+  NiceMock<Stats::MockIsolatedStatsStore> scope_;
+  std::unique_ptr<AsyncClientCacheSingleton> client_cache_singleton_;
+};
+
+TEST_F(AsyncClientCacheTest, Deduplication) {
+  Stats::IsolatedStoreImpl scope;
+  testing::InSequence s;
+
+  ::envoy::config::core::v3::GrpcService config;
+  config.mutable_google_grpc()->set_target_uri("dns://test01");
+  config.mutable_google_grpc()->set_credentials_factory_name("test_credential01");
+
+  expectClientCreation();
+  Grpc::RawAsyncClientSharedPtr test_client_01 =
+      client_cache_singleton_
+          ->getOrCreateAsyncClientCache(async_client_manager_, scope_, tls_, config)
```

Comment:
I think it would make the test more readable if you:
 * put the client-cache into a temp, factoring out all the calls to `client_cache_singleton_->getOrCreateAsyncClientCache(async_client_manager_, scope_, tls_, config)`
 * added one more EXPECT_EQ to prove that you can factor that out, e.g.
```
AsyncClientCacheSharedPtr client_cache = client_cache_singleton_
          ->getOrCreateAsyncClientCache(async_client_manager_, scope_, tls_, config);
EXEPCT_EQ(client_cache,  client_cache_singleton_
          ->getOrCreateAsyncClientCache(async_client_manager_, scope_, tls_, config));
```
then use client_cache below instead of spelling that all out.